pkg/csi/service/wcp/controller.go

```diff
@@ -1013,10 +1019,77 @@ func (c *controller) DeleteVolume(ctx context.Context, req *csi.DeleteVolumeRequ
+			volumeInfo, err := volumeInfoService.GetVolumeInfoForVolumeID(ctx, req.VolumeId)
+			switch {
+			case err == nil:
+				// Get K8s client for StoragePolicyUsage CR.
+				restConfig, err := config.GetConfig()
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to fetch kubernetes config. Error: %+v", err)
+				}
+				cnsOperatorClient, err := k8s.NewClientForGroup(ctx, restConfig, cnsoperatorv1alpha1.GroupName)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to create CNSOperator client. Error: %+v", err)
+				}
+
+				// Fetch StoragePolicyUsage instance for storageClass associated with the volume.
+				storagePolicyUsageInstanceName := volumeInfo.Spec.StorageClassName + "-" +
+					storagepolicyusagev1alpha1.NameSuffixForPVC
+				storagePolicyUsageInstance := &storagepolicyusagev1alpha1.StoragePolicyUsage{}
+				err = cnsOperatorClient.Get(ctx, apitypes.NamespacedName{
+					Namespace: volumeInfo.Spec.Namespace,
+					Name:      storagePolicyUsageInstanceName},
+					storagePolicyUsageInstance)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to fetch %s instance with name %q from supervisor namespace %q. Error: %+v",
+						storagepolicyusagev1alpha1.CRDSingular, storagePolicyUsageInstanceName,
+						volumeInfo.Spec.Namespace, err)
+				}
+
+				// Decrease the used capacity in StoragePolicyUsage instance as we are deleting the volume.
+				newUsedVal := storagePolicyUsageInstance.Status.ResourceTypeLevelQuotaUsage.Used.DeepCopy()
+				newUsedVal.Sub(*volumeInfo.Spec.Capacity)
+
+				// Patch the StoragePolicyUsage instance.
+				patch := map[string]interface{}{
+					"status": map[string]interface{}{
+						"quotaUsage": map[string]interface{}{
+							"used": newUsedVal,
+						},
+					},
+				}
+				patchBytes, err := json.Marshal(patch)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to create patch for StoragePolicyUsage instance. Error: %+v", err)
+				}
+				err = cnsOperatorClient.Patch(ctx, storagePolicyUsageInstance,
+					client.RawPatch(apitypes.MergePatchType, patchBytes))
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to patch storagePolicyUsage instance %q in namespace %q while deleting "+
+							"volume %q. Error: %+v", storagePolicyUsageInstance.Name, storagePolicyUsageInstance.Namespace,
+						req.VolumeId, err)
+				}
+
+				// Delete the CNSVolumeInfo instance for this volume.
+				err = volumeInfoService.DeleteVolumeInfo(ctx, req.VolumeId)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to delete cnsVolumeInfo CR for volume: %q. Error: %+v", req.VolumeId, err)
```

Comment:
If failed to delete cnsVolumeInfo CR (but actual volume is already deleted), when the delete volume operation retries, will the "used" filed in StoragePolicyCR decrease again?

Reply:
Please refer to the conversations above.